# HG changeset patch
# Parent 86fafd5b167d743c87c6b4f29f574687881fd53c
Introspect sizeof(suseconds_t) with CMake

And organize the top-level CMakeLists.txt a little bit better



diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2,8 +2,6 @@
 
 PROJECT(LIFXD C)
 
-SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${LIFXD_SOURCE_DIR}/CMakeScripts)
-
 SET(CPACK_PACKAGE_VERSION_MAJOR "0")
 SET(CPACK_PACKAGE_VERSION_MINOR "0")
 SET(CPACK_PACKAGE_VERSION_PATCH "1")
@@ -16,11 +14,20 @@
 MESSAGE(STATUS "Install prefix: ${CMAKE_INSTALL_PREFIX}")
 MESSAGE(STATUS "Source directory: ${LIFXD_SOURCE_DIR}")
 
+SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${LIFXD_SOURCE_DIR}/CMakeScripts)
+
+### Platform checks ############################################################
+
 # TODO: we need at least 2.0.19-stable because of the logging defines
 FIND_PACKAGE(Event2 REQUIRED COMPONENTS core)
 FIND_PACKAGE(Endian REQUIRED)
+INCLUDE(CompatTimeMonotonic)
 
-INCLUDE(CompatTimeMonotonic)
+INCLUDE(CheckTypeSize)
+CHECK_TYPE_SIZE(suseconds_t SUSECONDS_T_SIZE)
+ADD_DEFINITIONS("-DLIFXD_SUSECONDS_T_SIZE=${SUSECONDS_T_SIZE}")
+
+### Global definitions #########################################################
 
 SET(CMAKE_C_FLAGS "-pipe -Wextra -Wall -Wstrict-prototypes -std=c99")
 
diff --git a/core/log.c b/core/log.c
--- a/core/log.c
+++ b/core/log.c
@@ -60,7 +60,11 @@
     }
     // '2015-01-02T10:13:16.132222+00:00'
     snprintf(
+#if LIFXD_SUSECONDS_T_SIZE == 4
+        strbuf, bufsz, "%d-%02d-%02dT%02d:%02d:%02d.%d%c%02ld:%02ld",
+#else
         strbuf, bufsz, "%d-%02d-%02dT%02d:%02d:%02d.%ld%c%02ld:%02ld",
+#endif
         1900 + tm_now.tm_year, 1 + tm_now.tm_mon, tm_now.tm_mday,
         tm_now.tm_hour, tm_now.tm_min, tm_now.tm_sec,
         now.tv_usec, tm_now.tm_gmtoff >= 0 ? '+' : '-', // %+02ld doesn't work
